### PR TITLE
Variant encoder

### PR DESCRIPTION
**Pull request by talIguaz**

### CODE REVIEW

File: pkg/appender/appender.go [hunk 1]
```diff
@@ -217,6 +218,10 @@ func (mc *MetricsCache) Add(lset utils.LabelsIfc, t int64, v interface{}) (uint6
 
 	if !ok {
 		metric = &MetricState{Lset: lset, key: key, name: name, hash: hash}
+		// if the (first) value is not float, use variant encoding, TODO: test w schema
+		if _, ok := v.(float64); !ok {
```
Comment (imakhlin) [inline pkg/appender/appender.go]: Maybe you should take care of `NaN`. With `NaN` you can't determine the type, so it's better to skip it

File: pkg/aggregate/aggregate.go [hunk 2]
```diff
@@ -194,7 +194,10 @@ type AggregatesList []Aggregate
 
 // Append a value to all aggregates
 func (a AggregatesList) Aggregate(t int64, val interface{}) {
-	v := val.(float64)
+	v, ok := val.(float64)
```
Comment (imakhlin) [inline pkg/aggregate/aggregate.go]: If you want to skip the `variants` you should call aggregate at all

File: pkg/appender/store.go [hunk 3]
```diff
@@ -224,7 +223,7 @@ func (cs *chunkStore) Append(t int64, v interface{}) {
 }
 
 // Return current, previous, or create new  chunk based on sample time
-func (cs *chunkStore) chunkByTime(t int64) *attrAppender {
+func (cs *chunkStore) chunkByTime(t int64, isVar bool) *attrAppender {
```
Comment (imakhlin) [inline pkg/appender/store.go]: minor: use self explaining names

File: pkg/chunkenc/chunk.go [hunk 4]
```diff
@@ -44,6 +43,8 @@ func (e Encoding) String() string {
 		return "none"
 	case EncXOR:
 		return "XOR"
+	case EncVar:
```
Comment (imakhlin) [inline pkg/chunkenc/chunk.go]: cosmetic: `EncVariant`

File: pkg/chunkenc/vartype.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+type VarChunk struct {
+	logger logger.Logger
+
+	b       []byte
+	samples uint16
+	offset  int
+}
+
+// NewVarChunk returns a new chunk with variant encoding of the given size.
```
Comment (imakhlin) [inline pkg/chunkenc/vartype.go]: In comment you say "given size" please correct comment or code
Shouldn't chunk size be configurable?

File: pkg/chunkenc/vartype.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+type varAppender struct {
+	logger logger.Logger
+
+	c       *VarChunk
+	samples *uint16
+	t       int64
+}
+
+func (a *varAppender) Encoding() Encoding {
+	return a.Chunk().Encoding()
```
Comment (imakhlin) [inline pkg/chunkenc/vartype.go]: > a.c.Encoding()

File: pkg/chunkenc/vartype.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+		}
+
+		a.appendWithUint(t, varTypeFloat64, math.Float64bits(val))
+
+	case string:
+		val := []byte(v.(string))
+		a.appendWithValue(t, varTypeString, val)
+
+	default:
+		a.logger.Error("unsupported type %T!\n", vType)
```
Comment (imakhlin) [inline pkg/chunkenc/vartype.go]: > yields **syntax error: unexpected type, expecting expression**
you should use `v` instead

File: pkg/chunkenc/vartype_test.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	suite.Require().Nil(err)
+
+	list := []string{"abc", "", "123456"}
+	t0 := time.Now().UnixNano() / 1000
+
+	for i, s := range list {
+		t := t0 + int64(i*1000)
+		appender.Append(t, s)
+		b := chunk.Bytes()
+		fmt.Println(t, s, len(b))
```
Comment (imakhlin) [inline pkg/chunkenc/vartype_test.go]: Do you really need those prints during the test?

File: pkg/chunkenc/vartype_test.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/nuclio/zap"
+	"github.com/stretchr/testify/suite"
+)
+
+type testVarEncoderSuite struct {
+	suite.Suite
+}
+
+func (suite *testVarEncoderSuite) TestStringEnc() {
```
Comment (imakhlin) [inline pkg/chunkenc/vartype_test.go]: What about rest data types?
> 	varTypeNil     byte = 0
	varTypeBlob    byte = 1
	varTypeString  byte = 2
	varTypeBool    byte = 3
	varTypeFloat32 byte = 4
	varTypeFloat64 byte = 5
	varTypeInt8    byte = 8
	varTypeInt16   byte = 9
	varTypeInt32   byte = 10
	varTypeInt64   byte = 11

File: pkg/aggregate/aggregate.go (hunk 2 above)
Comment (talIguaz) [reply pkg/aggregate/aggregate.go]: It happens automatically when ingesting data. Now if you have one tsdb with 2 metrics one of them is normal (like cpu) the other is of variant type (like log).
you created the table with default aggregates that are being aggregated on ingest for every metric in the DB so we need to ignore non-floats.
At least until a better solution.

File: pkg/appender/appender.go (hunk 1 above)
Comment (talIguaz) [reply pkg/appender/appender.go]: Discussed in person

**Comment by talIguaz** [comment]
As discussed in person, will extend the support to all other types (other than `string`) in another PR

**Pull request merged**